File: leetcode/Tree/leetcode-95/95.go
package tree

// TreeNode Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func generateTrees(n int) []*TreeNode {
	res := []*TreeNode{}
	if n < 1 {
		return res
	}
	return helper(1, n)
}

func helper(start, end int) []*TreeNode {
	res := []*TreeNode{}

	if start > end {
		res = append(res, nil)
		return res
	}

	for i := start; i <= end; i++ {
		left := helper(start, i-1)
		right := helper(i+1, end)

		for _, l := range left {
			for _, r := range right {
				node := &TreeNode{i, nil, nil}
				node.Left = l
				node.Right = r
				res = append(res, node)
			}
		}
	}
	return res
}

File: leetcode/Tree/leetcode-95/95.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<TreeNode> generateTrees(int n) {
        List<TreeNode> res = new ArrayList<>();
        if (n == 0) {
            return res;
        }

        return helper(1, n);
    }

    private List<TreeNode> helper(int start, int end) {
        List<TreeNode> res = new ArrayList<>();
        if (start > end) {
            res.add(null);
            return res;
        }

        if (start == end) {
            res.add(new TreeNode(start));
            return res;
        }

        for (int i = start; i <= end; i++) {
            List<TreeNode> left = helper(start, i - 1);
            List<TreeNode> right = helper(i + 1, end);
            for (int m = 0; m < left.size(); m++) {  
                for (int n = 0; n < right.size(); n++) {
                    TreeNode root = new TreeNode(i);
                    root.left = left.get(m);
                    root.right = right.get(n);
                    res.add(root);
                }
            }
        }
        return res;
    }
}